### ISSUE TITLE
Sub-Orchestrations

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#15]
## Summary
This feature allows orchestrator functions to start and wait on sub-orchestrations as part of their logic. Any value returned by the sub-orchestration would be returned to the parent orchestration.

## Motivation
This is useful when you have a library of orchestrations and want to write an orchestration around them.

## Design Notes
The existing `DurableOrchestrationContext.CallFunctionAsync` API would be updated to allow the current orchestrator function to start and wait on other orchestrations, similar to how it calls activity functions.

This is supported by the Durable Task Framework today.  More information can be found [here](https://github.com/Azure/durabletask/wiki/Feature---Sub-orchestrations-%28aka-nested-orchestrations%29).

## Drawbacks
This design does not allow communication between the main orchestration and sub-orchestrations except via inputs and outputs. In that sense, a sub-orchestration is "embedded" into the main orchestration. Actor-style development, on the other hand, would require an ability to start a child-orchestration asynchronously, get back the instance ID, and send messages to it. A separate feature should be designed for this if desired.

**Comment by mmisztal1980** [comment]
+1 I've just encountered this, while implementing a 2-tier fan-out. Having the ability to call orchestration functions from orchestration functions would definitely be cool.

Is there an alternative available at the moment?

**Comment by cgillum** [comment]
There are ways to make this work, but some effort would be involved. Specifically, you could do the following:

1. Create an activity function which starts new orchestrations using `DurableOchestrationClient.StartNewAsync` and returns the instance ID.
2. Create another activity function which queries the status of existing orchestrations using `DurableOrchestrationClient.GetStatusAsync(instanceId)`
3. Have the parent orchestration trigger the child-orchestration via (1) and then wait for it to finish by calling (2) in a loop. This loop code would need to handle null responses and interpret them as "Pending".

So not a great experience right now.

**Comment by mmisztal1980** [comment]
I've encountered a problem while trying to implement 1). I'm getting an exception stating that my sub-orchestration func is expected to have an `[ActivityTrigger]`:

```
System.ArgumentException: 'No activity function named 'os-substations-data-import-orchestrator-func' with version '' is registered.'
```

**Comment by cgillum** [comment]
This is expected. `CallFunctionAsync` can only be used to call activity functions - i.e. those with `[ActivityTrigger]`. Once we implement the sub-orchestration work, you'll be able to call one orchestration from another directly.

Until then, the workaround is to create a generic activity function that binds to a `DurableOrchestrationClient` and calls `client.StartNewAsync`, and invoke that activity function from any orchestration code that wants to schedule a sub-orchestration.

Here is an example (warning, I haven't tested it yet):

```csharp
[FunctionName("StartSubOrchestration")]
public static async Task<string> Run(
    [ActivityTrigger] DurableActivityContext ctx,
    [OrchestrationClient] DurableOrchestrationClient starter,
    StartOrchestrationArgs args,
    TraceWriter log)
{
    // Function input comes from the request content.
    string instanceId = await starter.StartNewAsync(
        args.FunctionName,
        args.Input);

    log.Info($"Started orchestration with ID = '{instanceId}'.");
    return instanceId;
}
```

Note that `StartOrchestrationArgs` is a built-in type that you can use for this purpose.

**Comment by mmisztal1980** [comment]
I've implemented the trigger func as you've instructed, however, I'm still getting the same exception

**Comment by cgillum** [comment]
The "sub-orchestration" should be `[OrchestrationTrigger]`.  The exception message implies that you're trying to schedule an orchestration using `CallFunctionAsync`, which can only be used for `[ActivityTrigger]` functions.

A couple things to watch out for:
1. Make sure you spelled the name of the function correctly.
2. Make sure the error message you're seeing corresponds to the most recent invocation of your code (and that you're not seeing a replay of an older failure). You can clean out the data in your storage account if you want to make sure you're running in a clean state (look for queues and tables with "durablefunction" in the name - they get recreated automatically if they don't exist).

I'm doing some work which will hopefully improve the quality of these error messages to make it easier to find out what's wrong.

**Comment by mmisztal1980** [comment]
Here's my sub-orchestration trigger func:

```C#
 public static class SubStationsDataImportTriggerFunc
    {
        public const string Name = "os-substations-data-import-trigger-func";

        [FunctionName(Name)]
        public static async Task<string> Run([ActivityTrigger] DurableActivityContext context,
            [OrchestrationClient] DurableOrchestrationClient client,
            StartOrchestrationArgs args,
            TraceWriter log)
        {
            var orchestrationId = await client.StartNewAsync(SubstationsDataImportOrchestrationFunc.Name, args.Input);
           
            var station = args.Input as StationMetadata;

            log.Info($"Started subStation[{station.Station.Id}] data sync {orchestrationId}");

            return orchestrationId;
        }
    }
```

And here's the actual **sub-orchestration** func, as you can see, it's not doing anything yet, and has an `[OrchestrationTrigger]`:

```C#
 public static class SubstationsDataImportOrchestrationFunc
    {
        public const string Name = "os-substations-data-import-orchestrator-func";

        [FunctionName(Name)]
        public static async Task Run([OrchestrationTrigger] DurableOrchestrationContext context,
            TraceWriter log)
        {
            var station = context.GetInput<StationMetadata>();
        }
    }
```

**Comment by cgillum** [comment]
What you have is correct. The error message, however, is reporting something different - some *other* code is attempting to schedule an *activity trigger function* named `os-substations-data-import-orchestrator-func`.  Can you share the code which is calling `os-substations-data-import-trigger-func`?

**Comment by mmisztal1980** [comment]
Sure, here it is, the call is being made in the 2nd fan-out:

```C#
    public static class StationDataImportOrchestrationFunc
    {
        public const string Name = "os-station-data-import-orchestration-func";

        [FunctionName(Name)]
        public static async Task Run([OrchestrationTrigger] DurableOrchestrationContext context,
            TraceWriter log)
        {
            log.Info("Retrieving a list of devices from the repository");

            var stations = await context.CallFunctionAsync<Station[]>(StationListFunc.Name, null);

            Task<StationMetadata>[] tasks = new Task<StationMetadata>[stations.Length];
            for (var i = 0; i < stations.Length; i++)
            {
                tasks[i] = context.CallFunctionAsync<StationMetadata>(StationMetadataImportFunc.Name, stations[i]);
            }

            await Task.WhenAll(tasks);

            var metaData = tasks.Select(x => x.Result).ToArray();

            Task[] orchestrationTasks = new Task[stations.Length];
            for (var i = 0; i < stations.Length; i++)
            {
                orchestrationTasks[i] = context.CallFunctionAsync(SubstationsDataImportOrchestrationFunc.Name, metaData[i]);
            }

            await Task.WhenAll(orchestrationTasks);
        }       
    }
```

**Comment by cgillum** [comment]
Thanks. The problem is this line inside your second for loop:

```csharp
orchestrationTasks[i] = context.CallFunctionAsync(SubstationsDataImportOrchestrationFunc.Name, metaData[i]);
```

You're trying to call the orchestrator function using `CallFunctionAsync`, which isn't supported.  Instead, I think you want to change it to call the sub-station *trigger* func:

```csharp
orchestrationTasks[i] = context.CallFunctionAsync(SubStationsDataImportTriggerFunc.Name, metaData[i]);
```

**Comment by mmisztal1980** [comment]
oh crap, I can't believe I've missed this x(

**Comment by cgillum** [comment]
No worries. This is an easy mistake to make and this is great feedback for me that we need to improve our handling of cases like this (e.g. throw an exception in `CallFunctionAsync` if an invalid function name is provided - which I'm implementing now as we speak).
